docs/processes/quality-assurance/git-contribution-guide.md

```diff
@@ -168,6 +156,32 @@ git config --global core.editor vim
 git config --global core.editor "atom --wait"
 ```
 
+## Code Review Guidelines
+
+You are strongly encouraged to get your code reviewed by a reviewer as soon as
+there is any code to review, to get a second opinion on the chosen solution and
+implementation, and an extra pair of eyes looking for bugs, logic problems, or
+uncovered edge cases.
+
+### Draft Pull Request
+
+Draft Pull Requests allow a work in progress to receive early feedback. The developer creating the PR should, in the description, indicate their progress and any particular aspect they are looking for feedback on. When the PR is ready for final review, the developer should update the description and re-request reviews as required.
+
+Pull requests (draft and otherwise) are created from the GitHub website. Further information about draft pull requests, how to make them, and how to convert their status to ready for merging can be found on GitHub's [Introducing Draft Pull Requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/) blog.
+
+### Required Approvals
+
+Pull requests require **a minimum of two approvals**. The default approach is to
+choose a reviewer from your team for the first review. However, the reviewer may
+be from different team, for example a domain expert in Ruby or C++. Depending on
+the team size and dynamics, the required approvals can be higher or lower.
+Please try to have at least another person to look at the changes beside the
```

Comment:
I think rather than saying please try to have another person look at the changes, it should say both approvals should be people other than the author. Also clarifying who the second approval should be?

Reply:
eg. DL, PL or AL

Reply:
Thanks, @ginge270 

> I think rather than saying please try to have another person look at the changes, it should say both approvals should be people other than the author.

My intention is not to enforce a minimum of 2 approvals but to let teams decide to balance velocity and quality. For example, a small team might just need one or the delivery lead might not involve with much of the technical details. Furthermore, the system will not allow authors to approve themselves. Another intricacy is work done by a pair. Should the team allow it to expedite the approval process? I lean towards no but teams might have different opinions.

> Area Lead

This is an interesting one. I think approval should stay with the knowledge authority, not hierarchical authority. The process is to ensure we collect relevant insights and innovate, but less about asking for permission to execute. For example, some might not know about C++ or build systems so PRs around this area might not be suited to direct to them. It is good for upskilling as FYI but not from approval standpoint.

Reply:
I think this should be discussed further as my understanding from our QA meetings is not aligned with this.

An area lead or product lead may not have the knowledge authority, but QA is not just about the coding itself. Having someone to look at the PR to ensure that it meets the standards around readability, understandability, reusability - not just functionality and these are all transferable skills from other languages. Has the person who has submited the PR proven that this code does what it says it will. A peer or member of their team will be looking at it from a similar lens with the same context. 
 
 "Please try to have at least another person to look at the changes beside the author" implies that the system will allow a merge without approval or the author can approve themself - this is why I made the comment about being more explicit.

Reply:
Thank you for elaborating on your thoughts ❤️ I think we are aligned in principle. 

> knowledge authority, but QA is not just about the coding itself.

Yes, and knowledge authority is not about coding only.

> we are the ones most familiar with this vision and part of teaching this will be through feedback.

Yes totally, and I agree this is driven by knowledge, not authority.

> Having someone to look at the PR to ensure that it meets the standards around readability, understandability, reusability - not just functionality and these are all transferable skills from other languages...Does it make sense? A peer or member of their team will be looking at it from a similar lens with the same context.

This is a great point and I think it captures the intention of the approval process really well. We can include that in the doc. To be clear, I am fully supportive to expose the PR to more parties to invite diverse opinions, but not forcing approvals if it does not make sense. That needs to balance with velocity, everyone is working async and if we have a lot of required approvals, that would slow down the process and potentially affect people who propose changes. 

> "Please try to have at least another person to look at the changes beside the author" 

Sorry if it sounds confusing. I can try to reword it.

Reply:
I just thought I would add my untechnical opinion.
Getting two approvals is difficult, for this handbook we have had to contact each other numerous times just to have one approved review.

I understand the importance but with small teams this might be more difficult than we assume. Having code sit waiting for someone to approve it will drastically increase stress levels, create merging issues, and potentially reduce the actual care around quality - e.g. people approving anything.

Or are we assuming that by publishing the PR this is counted as 1 review?

Reply:
I have a suspicion practicality might force us to have a balance. Although... docs might be the kind of thing that you'd expect to end up with more PRs and less approvals needed, so maybe not representative. I think my tendency is to overload my coding commits/PRs so my gut feel on that might not be great either.

In my original vision (when we had different branch structures and PRs-getting-to-Doubtfire methods proposed) I was thinking the PRs would be much bigger chunks that "higher level" leads might be approving (with a much higher level, superficial kind of assessment than, say, @Tan usually does).

For instance, in my mind, this was where the "folder/branch for each project" structure came in - eg, a coder might have another reviewer approve their sub-branch to project branch merge, but then PRs to periodically merge changes to main would have a higher-level approver (so you'd need an area lead to look it over before it got into Thoth-Tech main, but there might have been many smaller steps that were peer-reviewed to get to the project branch before it was merged).

I now understand there's reasons for smaller PRs being approved & merged (including increasing potential for merge issues) but having removed the interim project branches from the flow, it does mean rather than fewer (more manageable) "check-point" PRs to assess (with larger, less manageable(?) code chunks), we may run into a pure PR quantity issue, as @talzid mentions.

We could have a "training period" where we force the extra approval to try to force quality standards 
until confidence is built with teams that they've got the hang of it and/or delegate this further down to delivery leads as they show they're confident in QA control.

- problem I see there is that this is the training period for everything and I can imagine this confidence might take a bit of the remaining trimester.
- of course, the other side of that coin is that if there's reason for that confidence to be shaky now, the chance of QA being followed without some approval gate-keeping may not be so good?

If, hypothetically, we think it's coming down to (A) Getting appreciable coding done (with 1 peer review only) or (B). Slowing progress to a trickle (but having more solid QA) could there be an option (C) whereby we let them go with 1 approval but do auditing instead and feed that back as issues? (I don't know much about issues - can they be raised post-merged-PR looking at prior changes?)

Reply:
@talzid Thank you for sharing your thoughts 🙏🏼 

> Getting two approvals is difficult

Yep. I share your concern. It's hard for small teams and/or depending on what stage our software product is. I would like to encourage teams to look at the suggestion and within their own capacity, reason about the appropriate approval process for them. 

Another thing is trying to balance individual workload with the expectation around the review cycle, (e.g. PR should receive a response in 24 hours). We can provide some more guidance on this one in follow-up PR.

> Or are we assuming that by publishing the PR this is counted as 1 review?

I don't think publishing the PR is counted as 1 review. A PR can have many reviewers but only receive approval from some. Or might be I misunderstood your question completely 🤔

Reply:
Thank you for your input @kats8 👍🏼 

>  I have suspicion practicality might force us to have a balance

Yes, it is. I would like to massage the message to give people that flexibility.

> We could have a "training period" where we force the extra approval to try to force quality standards

That's a great idea and in fact, most open-source teams work that way. Practically, I am not sure how we can do that given the limited time of the course and concentration of knowledge across products and technologies.

> (A) Getting appreciable coding done (with 1 peer review only) 
> (B) Slowing progress to a trickle (but having more solid QA)
> (C) whereby we let them go with 1 approval but do auditing instead and feed that back as issues

I think teams should go with 2 as recommended. Observe the quality metrics (bugs, incidents) and readjust. C is subjected to availability (when you said we, I assumed it is the Area Leads). The team can even do that themselves, for example, run a bug bash day at the end of the iteration.

> I don't know much about issues - can they be raised post-merged-PR looking at prior changes?

Yep, we can raise bug issues on the Trello board or PRs (if straight-forward and urgent) to address problems introduced by previous changes.

Reply:
I think teams should go with 2 as recommended. Observe the quality metrics (bugs, incidents) and readjust. C is subjected to availability (when you said we, I assumed it is the Area Leads). The team can even do that themselves, for example, run a bug bash day at the end of the iteration.

I agree with this outcome!

Reply:
Also, on the point that Katrina brought up - I think it should be clear in the guidelines that PRs should not be big pieces of work. They should be smaller chunks of functional work. As well as commit often

Reply:
> Also, on the point that Katrina brought up - I think it should be clear in the guidelines that PRs should not be big pieces of work. They should be smaller chunks of functional work. As well as commit often

In the spirit of iteration and small PR, I will create a follow-up PR to include this as a sub-section. Potentially include the content from the Git/GitHub training [slide](https://deakin365.sharepoint.com/:p:/r/sites/ThothTech2/_layouts/15/Doc.aspx?sourcedoc=%7BDE2764FD-E00C-42C5-BAA1-A2D7DDA4F444%7D&file=Git%20and%20GitHub%20Training.pptx&action=edit&mobileredirect=true). What do you think?

Reply:
If we are wanting to create some more work... we could turn the Github training session into .md and put it in training and link to that.

Reply:
Yep. It's on my list :)